### ISSUE TITLE
Reverse engineer Orbitrap peak quantification

### ISSUE THREAD

**Issue by oskarsari** [ultra-trace-systems/mascope#416]
Study how peak areas are calculated, and make sure we produce similar results in Mascope.

**Comment by ghost** [comment]
I found the latest `RawFileReader` DLLs, an official way to interact with Orbitrap *.raw files. I managed to import and use them in Python using `pythonnet` package. The investigation of *.raw files led to interesting results.

We don't need to precompute the m/z grid during the file conversion. We can read all mz and count arrays straight from the file. We can build the dataframe with one common m/z scale assigning corresponding m/z arrays from time spans to counts as indices. Such a dataframe facilitates writing data to the `zarr` file and this approach is much faster than our current one. Besides that, it naturally eliminates inconsistencies in peak shapes, shifts in peaks, and differences in peak heights noticed by me and Jussi in Mascope compared to Tracefinder. Some of them are shown in the picture below, the fitting will be discussed in further comments.

<img src="https://github.com/Karsa-Oy/Mascope/assets/162983707/6ecf5472-960d-4634-bd61-22bec7d14c5a" width=500 />

**Comment by ghost** [comment]
Tracefinder likely does not perform peak fitting. The internal Orbitrap device firmware calculates peak positions/centroids on the go and data is already available in the *.raw file. The algorithms they likely use in the Orbitrap firmware, "boxcar smoothing", "center of mass", and thresholding are basic and may skip overlapping or small peaks. However, we rarely observe overlapping peaks in Orbitrap.

I managed to reproduce the Orbitrap peak quantification with essentially the same output as in Tracefinder. Mascope gives almost the same result if it uses the raw spectrum extracted using `RawFileReader` DLLs.

However, there is a huge difference if we use our current routine (RawStreamer) to read raw files. Among issues compared to using RawFileReader DLLs: the raw peaks in Mascope are shifted, the intensity of raw peaks is way higher or way lower, some peaks are present in Mascope raw signal but not in obtained with RawFileReader DLLs (most of them are small and look like junk), some peaks with really high intensity were not fitted, as shown in the picture in my previous comment.

In conclusion, I think it would be beneficial to use the native `RawFileReader`.

<img src="https://github.com/Karsa-Oy/Mascope/assets/162983707/a79109fb-c87e-4c04-91b0-0ef8ffde4145" width=700 />

**Comment by ghost** [comment]
I also took a raw TOF signal and pushed it through a similar pipeline:
1. **Denoizing**. I compared Wavelet and Savitzky-Golay filtering. [Wavelet](https://pywavelets.readthedocs.io/en/latest/) denoising provided a much better result with no fine-tuning required.
2. **Background subtraction**. For baseline subtraction, the most promising is the Sensitive Nonlinear Iterative Peak ([SNIP](https://opg.optica.org/as/abstract.cfm?uri=as-62-1-91)) algorithm. But even though it provides fantastic results, it is computationally costly and requires further optimization.
3. **Boxcar filtering**. It is used to improve the signal-to-noise ratio, however, I didn't notice much improvement.
4. **Thresholding**. Estimating the threshold is the most challenging. First I tried the robust Median Absolute Deviation (MAD) method. The estimate was obviously too low, and the algorithm was too dependent on the chosen region. I ended up with a combination of Wavelets for Noise Thresholding and Stein's Unbiased Risk Estimator (SURE). The approach was fast, didn't require fine-tuning, and gave a fair estimation. 
5. **Peak fitting**. I used our common fitting routine but compared segmentations. Orbitrap-like segmentation is better at distinguishing peaks from noise. Generally, the result was the same as with a common TOF-like segmentation applied to a raw signal.

**My conclusion**: it's worth it only if we can optimize baseline subtraction.

In the figure below, Orbitrap-like segmentation was applied to a cleaned signal and TOF-like segmentation - to a raw signal.

<img src="https://github.com/Karsa-Oy/Mascope/assets/162983707/513ff6a6-3567-4a10-aa2c-afeb4ead9ba0" width=600/>

**Comment by ghost** [comment]
All the details are available [in the notebook](https://github.com/Karsa-Oy/Mascope/blob/feature/investigate-orbitrap-peak-fitting/notebooks/simulate_orbitrap_peak_fitting.ipynb).

**Comment by ghost** [comment]
And the comment about peak areas. I guess, the issue raised after Jussi mentioned that they are different in Mascope and Tracefinder. They are indeed different - because they are different things. In Mascope we calculate the peak area. Tracefinder integrates the change of the peak height with time. Jussi later explained that he didn't mean the area value but the peak height vs time trend. The trend is different because we have a peak split in the old Mascope version and the software fits two peaks instead of one. However, the issue does not persist in the latest Mascope.

<img src="https://github.com/Karsa-Oy/Mascope/assets/162983707/6d9419a7-aed5-45e1-9752-99726749a676" width=800/>
<img src="https://github.com/Karsa-Oy/Mascope/assets/162983707/3beb74e1-d15e-4cd5-836a-e66b8d3df764" width=500/>
